File: examples/go/push.go
// CPCP push example: write access. Stdlib only (net/http, encoding/json,
// crypto/rand).
//
// A PUSH names its intent before performing it: asking twice with the same
// name must not perform it twice. This example WRITES to the backend --
// point CPCP_URL at your own pod.
//
// Usage:
//
//	CPCP_URL=http://localhost:13002/_cpcp go run push.go [method] [params-json] [operation-id]
package main

import (
	"bytes"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"strings"
	"time"
)

func operationID() string {
	var b [8]byte
	if _, err := rand.Read(b[:]); err != nil {
		return "example-fallback"
	}
	return "example-" + hex.EncodeToString(b[:])
}

func push(baseURL, method string, params map[string]any, op string) (int, map[string]any) {
	if op == "" {
		op = operationID()
	}
	// Trailing slashes must not double the path: every client normalizes
	// the seam root before appending /rpc.
	baseURL = strings.TrimRight(baseURL, "/")
	body, _ := json.Marshal(map[string]any{
		"jsonrpc": "2.0", "id": 1, "method": method, "params": params, "operationId": op,
	})
	client := &http.Client{Timeout: 30 * time.Second}
	req, err := http.NewRequest("POST", baseURL+"/rpc", bytes.NewReader(body))
	if err != nil {
		return 0, map[string]any{"ok": false, "reason": "unreachable", "because": err.Error()}
	}
	req.Header.Set("Content-Type", "application/json")
	res, err := client.Do(req)
	if err != nil {
		return 0, map[string]any{"ok": false, "reason": "unreachable", "because": err.Error()}
	}
	defer res.Body.Close()
	var env map[string]any
	if err := json.NewDecoder(res.Body).Decode(&env); err != nil {
		return res.StatusCode, map[string]any{"ok": false, "reason": "cpcp_unparseable", "because": "no JSON body"}
	}
	return res.StatusCode, env
}

func main() {
	base := os.Getenv("CPCP_URL")
	if base == "" {
		base = "http://localhost:13002/_cpcp"
	}
	method := "note.create"
	if len(os.Args) > 1 {
		method = os.Args[1]
	}
	params := map[string]any{"title": "hello from cpcp", "body": "posted by the CPCP push example"}
	if len(os.Args) > 2 {
		params = map[string]any{}
		if err := json.Unmarshal([]byte(os.Args[2]), &params); err != nil {
			fmt.Fprintln(os.Stderr, "params must be JSON")
			os.Exit(2)
		}
	}
	op := ""
	if len(os.Args) > 3 {
		op = os.Args[3]
	}
	status, env := push(base, method, params, op)
	out, _ := json.MarshalIndent(map[string]any{"status": status, "envelope": env}, "", "  ")
	fmt.Println(string(out))
	if env["ok"] != true {
		os.Exit(1)
	}
}

File: examples/cpp/push.cpp
// CPCP push example: write access. C++17, POSIX sockets, no dependencies.
//
// A PUSH names its intent before performing it. This example WRITES to the
// backend -- point CPCP_URL at your own pod. Same minimal discipline as
// pull.cpp.
//
// Usage:
//   c++ -O2 -std=c++17 -o push push.cpp && CPCP_URL=http://localhost:13002/_cpcp ./push [title] [operation-id]
//
// Only plain http://host[:port]/path URLs (no TLS).
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <ctime>
#include <string>
#include <netdb.h>
#include <unistd.h>
#include <sys/socket.h>

namespace {

const char *kBase = "http://localhost:13002/_cpcp";
const size_t kCap = 256 * 1024;

bool split_url(const std::string &url, std::string &host, std::string &port, std::string &path) {
    std::string p = url;
    const std::string scheme = "http://";
    if (p.compare(0, scheme.size(), scheme) == 0) p = p.substr(scheme.size());
    size_t slash = p.find('/');
    std::string authority = slash == std::string::npos ? p : p.substr(0, slash);
    size_t colon = authority.find(':');
    if (colon == std::string::npos) {
        host = authority;
        port = "80";
    } else {
        host = authority.substr(0, colon);
        port = authority.substr(colon + 1);
    }
    std::string pp = slash == std::string::npos ? "/" : p.substr(slash);
    while (pp.size() > 1 && pp.back() == '/') pp.pop_back();
    const std::string rpc = "/rpc";
    path = (pp.size() >= rpc.size() && pp.compare(pp.size() - rpc.size(), rpc.size(), rpc) == 0)
        ? pp : pp + rpc;
    return !host.empty();
}

std::string escape_json(const std::string &in) {
    std::string out;
    for (char c : in) {
        if (c == '"') out += "\\\"";
        else if (c == '\\') out += "\\\\";
        else if (c == '\n') out += "\\n";
        else out += c;
    }
    return out;
}

int post(const std::string &base, const std::string &payload) {
    std::string host, port, path;
    if (!split_url(base, host, port, path)) {
        printf("{\"status\": 0, \"ok\": false, \"reason\": \"bad_url\"}\n");
        return 1;
    }
    std::string req = "POST " + path + " HTTP/1.1\r\nHost: " + host +
        "\r\nContent-Type: application/json\r\nContent-Length: " + std::to_string(payload.size()) +
        "\r\nConnection: close\r\n\r\n" + payload;

    addrinfo hints{};
    hints.ai_socktype = SOCK_STREAM;
    addrinfo *addrs = nullptr;
    if (getaddrinfo(host.c_str(), port.c_str(), &hints, &addrs) != 0) {
        printf("{\"status\": 0, \"ok\": false, \"reason\": \"unreachable\"}\n");
        return 1;
    }
    int fd = -1;
    for (addrinfo *a = addrs; a; a = a->ai_next) {
        fd = socket(a->ai_family, a->ai_socktype, a->ai_protocol);
        if (fd < 0) continue;
        if (connect(fd, a->ai_addr, a->ai_addrlen) == 0) break;
        close(fd);
        fd = -1;
    }
    freeaddrinfo(addrs);
    if (fd < 0) {
        printf("{\"status\": 0, \"ok\": false, \"reason\": \"unreachable\"}\n");
        return 1;
    }
    size_t off = 0;
    while (off < req.size()) {
        ssize_t n = write(fd, req.data() + off, req.size() - off);
        if (n <= 0) break;
        off += (size_t)n;
    }
    std::string buf;
    buf.resize(kCap);
    size_t total = 0;
    ssize_t n;
    while (total < kCap - 1 && (n = read(fd, &buf[total], kCap - 1 - total)) > 0)
        total += (size_t)n;
    close(fd);
    buf.resize(total);
    int status = 0;
    sscanf(buf.c_str(), "HTTP/%*s %d", &status);
    size_t hdr = buf.find("\r\n\r\n");
    std::string body = hdr == std::string::npos ? buf : buf.substr(hdr + 4);
    bool ok = body.find("\"ok\":true") != std::string::npos ||
              body.find("\"ok\": true") != std::string::npos;
    printf("{\"status\": %d, \"body\": %s}\n", status, body.empty() ? "{}" : body.c_str());
    return ok ? 0 : 1;
}

}  // namespace

int main(int argc, char **argv) {
    const char *env = getenv("CPCP_URL");
    std::string base = (env && *env) ? env : kBase;
    std::string title = argc > 1 ? argv[1] : "hello from cpcp";
    char opid[64];
    if (argc > 2) {
        snprintf(opid, sizeof opid, "%s", argv[2]);
    } else {
        snprintf(opid, sizeof opid, "example-%lx%lx",
                 (unsigned long)time(nullptr), (unsigned long)getpid());
    }
    std::string payload =
        "{\"jsonrpc\":\"2.0\",\"id\":1,\"method\":\"note.create\","
        "\"params\":{\"title\":\"" + escape_json(title) +
        "\",\"body\":\"posted by the CPCP push example\"},"
        "\"operationId\":\"" + std::string(opid) + "\"}";
    return post(base, payload);
}
